modules/sitemaps/sitemap-buffer.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+/**
+ * A buffer for constructing sitemap xml files.
+ *
+ * Models a list of strings such that
+ *
+ * 1. the list must have a bounded number of entries,
+ * 2. the concatenation of the strings must have bounded
+ *      length (including some header and footer strings), and
+ * 3. each item has a timestamp, and we need to keep track
+ *      of the latest timestamp of the items in the list.
+ *
+ * Sitemaps (per the protocol) are essentially lists of XML fragments;
+ * lists which are subject to size constraints. This class abstracts
+ * the details of checking these constraints.
+ */
+class Jetpack_Sitemap_Buffer {
+	private $item_capacity;
+	private $byte_capacity;
+	private $footer_text;
+	private $buffer;
+	private $is_full_flag;  // True if we've tried to add something and failed.
+	private $timestamp;     // 'YYYY-MM-DD hh:mm:ss'
+
+	/**
+	 * Construct a new Jetpack_Sitemap_Buffer.
+	 *
+	 * @param int $item_limit The maximum size of the buffer in items.
+	 * @param int $byte_limit The maximum size of the buffer in bytes.
+	 * @param string $header The string to prepend to the entire buffer.
+	 * @param string $footer The string to append to the entire buffer.
+	 * @param string $time The initial timestamp of the buffer, in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	public function __construct(
+		$item_limit = 50000,    // 50k
+		$byte_limit = 10485760, // 10MB
+		$header = '',
+		$footer = '',
+		$time
+	) {
+		$this->item_capacity = $item_limit;
+		$this->byte_capacity = $byte_limit - mb_strlen($header) - mb_strlen($footer);
```

Comment:
On further inspection `mb_strlen` was the wrong thing to use here anyway- it specifically counts characters instead of bytes, which is what we need. I changed this to use just `strlen` instead, although reading some documentation I'm concerned that this will not do the right thing either as some hosts alias `strlen` to `mb_strlen`.